events/events.go

```diff
@@ -31,9 +35,83 @@ type EventListener interface {
 	HandleEvent(event Event)
 }
 
+// EventEmitter describes an interface that can be used to emit events
+type EventEmitter interface {
+	EmitEvent(Event)
```

Comment:
I wanted to be very specific here with the function names as they become part of the public interface for structs that embed the structs from below (which is the intended way of using the structs from below). Emit might not be evident enough that it is about emitting an event to its listeners depending on the place where these structs get embedded.